yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala

```diff
@@ -223,4 +225,43 @@ object YarnSparkHadoopUtil {
     )
   }
 
+  /**
+   * Expand environment variable using Yarn API.
+   * If environment.$$() is implemented, return the result of it.
+   * Otherwise, return the result of environment.$()
+   * Note: $$() is added in Hadoop 2.4.
+   */
+  def expandEnvironment(environment: Environment): String = {
+    var result = environment.$()
+
+    // We use reflection in order not to fail building with Hadoop 2.3 or before.
+    val clazz = classOf[Environment]
```

Comment:
This is probably not a big deal, since this is not in any hot path, but it seems wasteful to look for the right method every time this method is called instead of caching the method to be called. I'd do this:

    private lazy val expandMethod = Try(classOf[Environment].getMethod("$$"))
      .getOrElse(classOf[Environment].getMethod("$"))

Then your method becomes just:

    def expandEnvironment(environment: Environment): String = \
      expandMethod.invoke(environment).asInstanceOf[String]